### PR TITLE
feat: Add OS detection hook and update search dialog

### LINKED ISSUE #1758: Bug: User search shortcut shown in UI is macOS specific
### Describe the bug

The user search shortcut shown in the UI is macOS specific. CMD + K. The command key does not exist on other operating systems (OSes).

https://github.com/open-sauced/app/blob/e895f621df683e29ed0864ce18fb9dd42828a93e/components/organisms/SearchDialog/search-dialog.tsx#L96-L98

https://github.com/open-sauced/app/blob/e895f621df683e29ed0864ce18fb9dd42828a93e/components/organisms/SearchDialog/search-dialog.tsx#L140-L142

### Steps to reproduce

If you go to https://beta.insights.opensauced.pizza/hub/insights and search for a user, you can use CMD + K (macOS) or CTRL + K (other OSes). This works well, but the shortcut shown in the UI is only for macOS.

The shortcut should show CTRL + K or whatever we deem a shorter hand of CTRL + K.

CleanShot 2023-09-27 at 11 17 47

### Browsers

_No response_

### Additional context (Is this in dev or production?)

dev (feature hasn't rolled out to production yet)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by FrancoLab**
## Description
This PR addresses issue #1758 and introduces the following changes:

- Added a new file ﻿useOSDetection.ts which contains a custom hook for detecting the user's operating system.
- Updated the file ﻿search-dialog.tsx by importing and using the ﻿useOSDetection hook.
- Made necessary UI changes to handle OS-specific text in the search dialog.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [x] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [x] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Fixes #1758 

## Mobile & Desktop Screenshots/Recordings

!MacOS-Search
!MacOS-Dialog
MacOS-Search
MacOS-Dialog

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?
No

## [optional] What gif best describes this PR or how it makes you feel?
XD

### CODE REVIEW

**Review by jpmcb — COMMENTED** [review]
Very nice! Code wise, this looks solid. Thanks for tackling this.

The styling looks abit cramped to me?

Screenshot 2023-09-27 at 3 07 59 PM

I wonder if this also deserves it's own styling. Curious what others think?

File: lib/hooks/useOSDetection.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    const macOSPatterns = /Mac|Macintosh|MacIntel|Mac OS X/;
+
+    // Initial check using userAgent
+    const userAgentCheck = macOSPatterns.test(window.navigator.userAgent);
+
+    setIsMac(userAgentCheck);
+  }, []);
+
+  return isMac;
+}
```
Comment (nickytonline) [inline lib/hooks/useOSDetection.ts:17]: The regular expression doesn't need to be created in the hook. The hook is called `useOSDetection`, but what's it's doing is return whether or not it's macOS.

As mentioned in the issue I opened, I'm not sure it if really needs to be a React hook. It could be a regular function.

```suggestion
// Constant pattern to test against
const MACOS_PATTERNS = /Mac|Macintosh|MacIntel|Mac OS X/;
    
function useIsMac(): boolean {
  const [isMac, setIsMac] = useState<boolean>(false);

  useEffect(() => {
    // Initial check using userAgent
    const userAgentCheck = macOSPatterns.test(window.navigator.userAgent);

    setIsMac(userAgentCheck);
  }, []);

  return isMac;
}
```

**Review by nickytonline — COMMENTED** [review]
Thanks for working on this @FrancoLab!

**Comment by nickytonline** [comment]
> Very nice! Code wise, this looks solid. Thanks for tackling this.
> 
> The styling looks abit cramped to me?
> 
> Screenshot 2023-09-27 at 3 07 59 PM
> 
> I wonder if this also deserves it's own styling. Curious what others think?

I agree that it's a little cramped.

**Comment by FrancoLab** [comment]
> Very nice! Code wise, this looks solid. Thanks for tackling this.
> 
> The styling looks abit cramped to me?
> 
> Screenshot 2023-09-27 at 3 07 59 PM
> 
> I wonder if this also deserves it's own styling. Curious what others think?

Thank you for the comments! I can definitely work on updating the styling a bit and give it more room to breath.

File: lib/hooks/useOSDetection.ts (hunk 1 above)
Comment (FrancoLab) [reply lib/hooks/useOSDetection.ts:17]: Thank you, @nickytonline. I initially had the code as a function, but I ran into a scenario where the function needed to be called within a useEffect in the component, and the component was responsible for handling the state. To address this, I decided to refactor it into a custom hook, which abstracts away the complexity and only returns the value that the component requires. This way, the code becomes more modular and reusable.

Do you have any thoughts on this?

_PS: Looking at your suggestion above and also implementing styling fixes/updates._

File: components/organisms/SearchDialog/search-dialog.tsx [hunk 2]
```diff
@@ -23,6 +24,7 @@ const SearchDialog = () => {
   const setOpenSearch = store((state) => state.setOpenSearch);
   const debouncedSearchTerm = useDebounceTerm(searchTerm, 300);
   const [searchResult, setSearchResult] = useState<{ data: DbUserSearch[]; meta: {} }>();
+  const isMac = useOSDetection();
```
Comment (nickytonline) [inline components/organisms/SearchDialog/search-dialog.tsx:27]: In regards to your question about it being a hook or just a non-hook function, it would fail when rendering server-side with `ReferenceError: window is not defined`, so the custom hook works.

File: lib/hooks/useOSDetection.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+function useOSDetection(): boolean {
+  const [isMac, setIsMac] = useState<boolean>(false);
+
+  useEffect(() => {
+    // Constant pattern to test against
+    const macOSPatterns = /Mac|Macintosh|MacIntel|Mac OS X/;
+
+    // Initial check using userAgent
+    const userAgentCheck = macOSPatterns.test(window.navigator.userAgent);
```
Comment (nickytonline) [inline lib/hooks/useOSDetection.ts:11]: Prefer all caps for constans.

```suggestion
const MACOS_USER_AGENT_PART = /Mac|Macintosh|MacIntel|Mac OS X/;

function useOSDetection(): boolean {
  const [isMac, setIsMac] = useState<boolean>(false);

  useEffect(() => {

    // Initial check using userAgent
    const userAgentCheck = MACOS_USER_AGENT_PART.test(window.navigator.userAgent);
```

**Review by nickytonline — COMMENTED** [review]
Left some feedback on your question @FrancoLab

File: lib/hooks/useOSDetection.ts (hunk 3 above)
Comment (FrancoLab) [reply lib/hooks/useOSDetection.ts:11]: Awesome thanks. Making the updates as we speak. For the styling, do you suggest we lower the font-size and increase padding? I tried and even to add `px-1` we would have to drop the font-size to `text-[10px]`. Feels a bit small.

Comment (nickytonline) [reply lib/hooks/useOSDetection.ts:11]: I'll let you know @isabensusan chime in.

Comment (FrancoLab) [reply lib/hooks/useOSDetection.ts:11]: I can make it `text-[12px]` `px-1` and it looks pretty great. Would need to increase the width of the search box a bit to `w-56`.
Screenshot 2023-09-29 at 23 24 12
Screenshot 2023-09-29 at 23 29 52

Comment (brandonroberts) [reply lib/hooks/useOSDetection.ts:11]: @FrancoLab that seems reasonable to go with. We'll revisit if it needs adjustments in the future

Comment (FrancoLab) [reply lib/hooks/useOSDetection.ts:11]: Thanks @brandonroberts . Will implement these changes 💪

Comment (isabensusan) [reply lib/hooks/useOSDetection.ts:11]: lgtm! thanks 😄

Comment (nickytonline) [reply lib/hooks/useOSDetection.ts:11]: I just realized in my refactor suggestions, that I forgot to rename the hook to, `function useIsMacOS()`. The code suggestion above has been updated @FrancoLab.

Comment (FrancoLab) [reply lib/hooks/useOSDetection.ts]: Thanks everyone. New updates pushed. Ready for review.

File: components/organisms/SearchDialog/search-dialog.tsx [hunk 4]
```diff
@@ -1,17 +1,19 @@
...
 import Avatar from "components/atoms/Avatar/avatar";
 import { ScrollArea } from "components/atoms/ScrollArea/scroll-area";
-import useLockBody from "lib/hooks/useLockBody";
+import Text from "components/atoms/Typography/text";
 import { getAvatarByUsername } from "lib/utils/github";
 import { searchUsers } from "lib/hooks/search-users";
+import store from "lib/store";
 import useDebounceTerm from "lib/hooks/useDebounceTerm";
+import useIsMacOS from "lib/hooks/useIsMacOS";
+import useLockBody from "lib/hooks/useLockBody";
```
Comment (nickytonline) [inline components/organisms/SearchDialog/search-dialog.tsx:16]: It looks like all the imports were rearranged? Was that eslint/prettier, or is that an editor extension?

If it's not eslint/prettier, revert those changes as they are unrelated to the PR.

**Review by nickytonline — COMMENTED** [review]
This looks great @FrancoLab! Just left a question for you.

File: components/organisms/SearchDialog/search-dialog.tsx (hunk 4 above)
Comment (FrancoLab) [reply components/organisms/SearchDialog/search-dialog.tsx:16]: Will implement this fix tonight 🔥

**Review by nickytonline — APPROVED** [review]
Thanks for making the changes @FrancoLab! Nice work! 🚢

**Review by brandonroberts — APPROVED** [review]
